core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/CongestionControl.scala

```diff
@@ -0,0 +1,90 @@
+package edu.uci.ics.amber.engine.architecture.messaginglayer
+
+import edu.uci.ics.amber.engine.architecture.messaginglayer.CongestionControl.{
+  maxWindowSize,
+  minWindowSize,
+  timeGap
+}
+import edu.uci.ics.amber.engine.architecture.messaginglayer.NetworkSenderActor.NetworkMessage
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.duration.{DurationInt, FiniteDuration}
+
+class CongestionControl {
+
+  // ack should be received within 3s,
+  // otherwise, network congestion occurs.
+  final val ackTimeLimit = 3000
+
+  // if the ack for a message is not received after 60s,
+  // we trigger the resending logic.
+  // Note that the resend is not guaranteed to happen
+  // after sending the message for 60s
+  final val resendTimeLimit = 60000 // 60s
+
+  // slow start threshold
+  // after windowSize > ssThreshold,
+  // we increment windowSize by one every time,
+  // otherwise, we multiply windowSize by 2.
+  private var ssThreshold = 16
+
+  // initial window size = 1
+  // it represents how many messages can be sent concurrently
+  private var windowSize = 1
+
+  private val toBeSent = new mutable.Queue[NetworkMessage]
+  private val messageBuffer = new ArrayBuffer[NetworkMessage]()
+  private val inTransit = new mutable.LongMap[NetworkMessage]()
+  private val sentTime = new mutable.TreeMap[Long, Long]()
+
+  def canSend(data: NetworkMessage): Boolean = {
+    if (inTransit.size < windowSize) {
+      true
```

Comment:
put a comment to say that toBeSent buffer is always empty whenever inTransit.size < windowSize